File: src/leetcode/second/Code_208_Trie.java
package leetcode.second;

/**
 * trie
 * 
 * @since 2020-8-14 Friday 15:04 - 15:42
 */
public class Code_208_Trie {
    public static void main(final String[] args) {
        final Trie trie = new Trie();
        trie.insert("length");
        trie.insert("apple");
        System.out.println(trie.search("length"));
        System.out.println(trie.search("lengthh"));
        System.out.println(trie.startsWith("len"));
        System.out.println(trie.startsWith("leh"));
        System.out.println(trie.search("app"));
        trie.insert("app");
        System.out.println(trie.search("app"));
    }
}

class Trie {
    private final TrieNode node;

    /** Initialize your data structure here. */
    public Trie() {
        node = new TrieNode();
    }
    
    /** Inserts a word into the trie. */
    public void insert(final String word) {
        TrieNode n = node;
        for (int i = 0; i < word.length(); i++) {
            final char c = word.charAt(i);
            if (!n.containsKey(c)) {
                n = n.put(c);
            } else {
                n = n.get(c);
            }
        }
        n.setEnd(true);
    }
    
    /** Returns if the word is in the trie. */
    public boolean search(final String word) {
        TrieNode n = searchPrefix(word);
        return n == null ? false : n.isEnd();
    }
    
    /** Returns if there is any word in the trie that starts with the given prefix. */
    public boolean startsWith(final String prefix) {
        return searchPrefix(prefix) != null;
    }

    private TrieNode searchPrefix(final String prefix) {
        TrieNode n = node;
        for (int i = 0; i < prefix.length(); i++) {
            final char c = prefix.charAt(i);
            if (!n.containsKey(c)) {
                return null;
            } else {
                n = n.get(c);
            }
        }
        return n;
    }

    private static final class TrieNode {
        private final TrieNode[] links;
        private boolean isEnd;

        TrieNode() {
            links = new TrieNode[26];
        }

        TrieNode put(final char c) {
            final TrieNode n = new TrieNode();
            links[c - 'a'] = n;
            return n;
        }

        TrieNode get(final char c) {
            return links[c - 'a'];
        }

        boolean containsKey(final char c) {
            return links[c - 'a'] != null;
        }

        boolean isEnd() {
            return isEnd;
        }

        void setEnd(final boolean end) {
            isEnd = end;
        }
    }
}

File: src/c-leetcode/first/Code_208_Trie.c
/* trie

   @since 2020-9-10 Thursday
*/

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

typedef struct trie
{
    bool isEnd;
    struct trie *next[26];
} Trie;

/** Initialize your data structure here. */

Trie *trieCreate()
{
    Trie *res = calloc(1, sizeof(Trie));
    res->isEnd = false;
    return res;
}

/** Inserts a word into the trie. */
void trieInsert(Trie *obj, char *word)
{
    Trie *node = obj;
    for (; *word; word++)
    {
        if (node->next[*word - 'a'] == 0)
            node->next[*word - 'a'] = trieCreate();
        node = node->next[*word - 'a'];
    }
    node->isEnd = true;
}

/** Returns if the word is in the trie. */
bool trieSearch(Trie *obj, char *word)
{
    Trie *node = obj;
    for (; *word; word++)
    {
        if (node->next[*word - 'a'] == 0)
            return false;
        else
            node = node->next[*word - 'a'];
    }
    return node->isEnd;
}

/** Returns if there is any word in the trie that starts with the given prefix. */
bool trieStartsWith(Trie *obj, char *prefix)
{
    Trie *node = obj;
    for (; *prefix; prefix++)
    {
        if (node->next[*prefix - 'a'] == 0)
            return false;
        else
            node = node->next[*prefix - 'a'];
    }
    return true;
}

void trieFree(Trie *obj)
{
    free(obj);
}

int main(int argc, char const *argv[])
{
    // char *word = "word";
    char *word = "word";
    char *search = "wo";
    char *prefix = "w";
    Trie *obj = trieCreate();
    trieInsert(obj, word);
    bool param_2 = trieSearch(obj, search);
    printf("res p2: %d'\n", param_2);
    bool param_3 = trieStartsWith(obj, prefix);
    printf("res p3: %d'\n", param_3);
    trieFree(obj);
}
